src/backend/optimizer/plan/setrefs.c

```diff
@@ -396,23 +396,32 @@ set_plan_references(PlannerInfo *root, Plan *plan)
 	 */
 	add_rtes_to_flat_rtable(root, false);
 
+	int	glen = list_length(glob->finalrtable);
 	/*
 	 * Adjust RT indexes of PlanRowMarks and add to final rowmarks list
 	 */
 	foreach(lc, root->rowMarks)
 	{
 		PlanRowMark *rc = lfirst_node(PlanRowMark, lc);
-		PlanRowMark *newrc;
 
-		/* flat copy is enough since all fields are scalars */
-		newrc = (PlanRowMark *) palloc(sizeof(PlanRowMark));
-		memcpy(newrc, rc, sizeof(PlanRowMark));
+		if (plan->type != T_ShareInputScan || !rc->isIndexesAdjusted) 
```

Comment:
The solution does not apply to the root cause, while solving the problem. Have you answered the question, why double adjustment
for query like
```
 WITH cte AS (
select es1.i from es1 join es2 on es2.i = es1.i for update  
)
SELECT * FROM cte UNION ALL SELECT * FROM cte
```
works well, but for the query with update
```
WITH cte AS (
UPDATE es1 SET i = es1.i FROM es2 WHERE es2.i = es1.i RETURNING *
)
SELECT * FROM cte UNION ALL SELECT * FROM cte;
```
does not?

The fact is that for UPDATE/DELETE [we have adjustment of base rowmark](https://github.com/arenadata/gpdb/blob/adb-7.1.0/src/backend/optimizer/plan/setrefs.c#L1277-L1282) from  `root->rowMarks`. 
How the things work for the second query:
1. we get to setrefs.c:403  with rc->rti = 2 and rtooffset=4, newrc is adjusted, rc is not
2. we get to setrefs.c: 1281 with rc->rti = 2 and rtoofset = 4, rc is adjusted.
3. due to shared scan we get to setrefs.c:403 with rc->rti = 6 and rtoofset =7, and newrc is adjusted = 13.

For the first query works great, because there is not 2nd step, rc->rti does not change and adjustment does not occur.

Here is my remarks for the solution:
1. Can we avoid extending the PlanRowMark node with a new field?
2. Probably we can achieve the same valid behaviour as for common select for update case?

Reply:
>    Can we avoid extending the PlanRowMark node with a new field?
Provided commit solves the problem without a new field. However, there is a tradeoff between memory and CPU. In this case we exchange 1 byte (4 bytes aligned) for the O(N) list lookup. This particular list is not going to be long, so the difference won't be noticeable, but in general I'd prefer the first solution with a flag.

>    Probably we can achieve the same valid behaviour as for common select for update case?
Update makes ModifyTable node and its references accessed by the executor, so should be adjusted . Select for update makes SubqueryScan and do not adjust itself separately. In fact, select for update differ from simple select only in lock types.
I am afraid it's  possible only with some global refactoring.

>    Do you think that its valid to prohibit multiple adjustment and multiple identical PlanRowMarks in finalrowmarks list? Even for select for update case?
There is a single rowmark, shared by two ModifyTable nodes in this case, that's why the problem arises. So we just putting this rowmark to the global finalrowmarks list.

Reply:
> This particular list is not going to be long, so the difference won't be noticeable, but in general I'd prefer the first solution with a flag

I agree that adding a flag is more optimal. However, it's Postgres code , and extending of its structs should be a last resort. It also brings some maintenance complexity. Moreover such queries are not typical load for the GPDB, and for smaller lists the performance impact of iterating through the list might be negligible, thus maintaining simplicity in design.

>There is a single rowmark, shared by two ModifyTable nodes in this case, that's why the problem arises. So we just putting this rowmark to the global finalrowmarks list.

SELECT for UPDATE also uses in this case the same rowmarks for each cte refernce. Therefore for query 
```
 WITH cte AS (
select es1.i from es1 join es2 on es2.i = es1.i for update  
)
SELECT * FROM cte UNION ALL SELECT * FROM cte
```
we have 4 rowmarks in `global->finalrowmark` list. My question was: is it ok to duplicate rowmarks in `global->finalrowmark` list?

Reply:
> My question was: is it ok to duplicate rowmarks in global->finalrowmark list? Is it valid to have only one entry for each rowmark?

If we look how global->finalrowmarks used we can see that it needed to put into resulting plan:
[assign result->rowMarks](https://github.com/arenadata/gpdb/blob/f98bb29eb0a5421c5500f649773d2abf90289514/src/backend/optimizer/plan/planner.c#L755)

rowMarks is used (iterated) in:

- [lookup](https://github.com/arenadata/gpdb/blob/f98bb29eb0a5421c5500f649773d2abf90289514/src/backend/executor/execMain.c#L1611)
- [build the ExecRowMark array](https://github.com/arenadata/gpdb/blob/f98bb29eb0a5421c5500f649773d2abf90289514/src/backend/executor/execMain.c#L1778)

and a few more places where it's not iterated.

In it's turn ExecRowMark is used in:
- [ExecFindRowMark](https://github.com/arenadata/gpdb/blob/f98bb29eb0a5421c5500f649773d2abf90289514/src/backend/executor/execMain.c#L3409) where it's accessed by the index 

So I think it's safe both to duplicate rowmarks and reuse rowmark entries as long as indexes are properly set.

Reply:
> So I think it's safe both to duplicate rowmarks and reuse rowmark entries as long as indexes are properly set.

And now you you are not duplicating. So, can you please answer the question:
Is it valid to have only one entry for each rowmark?

Reply:
I think yes.